test/py/test_migration.py

```diff
@@ -204,26 +218,40 @@ def test_migration_shortest_state_transition_paths():
         VFIO_USER_DEVICE_STATE_PRE_COPY_P2P: set()
     }
 
+    # states (vertices)
+    V = E.keys()
+
     # "saving states" which cannot be internal arcs
-    S = {VFIO_USER_DEVICE_STATE_PRE_COPY, VFIO_USER_DEVICE_STATE_STOP_COPY}
+    saving_states = {VFIO_USER_DEVICE_STATE_PRE_COPY,
+                     VFIO_USER_DEVICE_STATE_STOP_COPY}
 
+    # Consider each vertex in turn to be the start state, that is, the state
+    # we are transitioning from.
     for source in V:
         back = {v: None for v in V}
```

Comment:
Backpointers for each node (the previous node in the shortest path from the start to the node). We use this dictionary to retrieve the shortest paths to compare with the server's ones.

Suppose the shortest path from a to c was `a -> b -> c`, then `back[b] == a`.

Reply:
The implementation isn't sufficiently explained, the fact that BFS isn't enough because this isn't a textbook implementation. I would suggest refactorin and explaining further, e.g. the `while` loop at the beggining could be made a function where the input and output can be documented, etc.

Reply:
Apart from the condition that we can't go through a saving state, it is a textbook implementation. However I do agree that having this condition embedded within the for loop is unnecessarily confusing (and also inefficient!) so I've moved it out of the for loop and documented exactly what it's doing and why it's there. Is this any better? @tmakatos